src/resource/package_builder.rs

```diff
@@ -0,0 +1,423 @@
+use std::collections::{HashMap, HashSet};
+use std::fs::File;
+use std::io;
+use std::io::{Read, Seek, Write};
+use std::path::{Path, PathBuf};
+
+use binrw::BinWrite;
+use binrw::io::Cursor;
+use thiserror::Error;
+
+use crate::resource::resource_package::{ChunkType, PackageHeader, PackageMetadata, PackageOffsetInfo, ResourceHeader, ResourcePackage, ResourceReferenceCountAndFlags, ResourceReferenceFlagsV2};
+use crate::resource::runtime_resource_id::RuntimeResourceID;
+
+pub enum PackageResourceBlob {
+    FromDisk(PathBuf),
+    FromMemory(Vec<u8>),
+}
+
+/// A builder for creating a resource within a ResourcePackage
+pub struct PackageResourceBuilder {
+    rrid: RuntimeResourceID,
+    blob: PackageResourceBlob,
+    resource_type: String,
+    data_size: u32,
+    system_memory_requirement: u32,
+    video_memory_requirement: u32,
+    references: HashMap<RuntimeResourceID, ResourceReferenceFlagsV2>,
```

Comment:
Ah ok. I had it as a vector initially but changed it to map to make de-duplication easier. Should I still add some deduplication logic or can there be duplicates? Also, does the order of the resources in the package matter?

Reply:
The only example of duplicate dependencies I've seen was with invalid `00FFFFFFFFFFFFFF` rrids. 
We can't remove duplicates, if someone references an index that we removed it would crash.

The official rpkg files are generated using a BFS on dependencies of the resourceIDs in a packagedefs file. This is how IOI makes sure that it only packs used resources. 

RPKG-Tool just adds resources unordered when it can't find a meta file. So the order of the resources probably doesn't matter to the game.